### ISSUE TITLE
[doobie][scala3] ZEnvironment missings layer

### ISSUE THREAD

**Issue by ex0ns** [gaelrenoux/tranzactio#44]
With the following code (scala-cli):

```scala
//> using scala "3.2.1"
//> using lib "dev.zio::zio:2.0.5"
//> using lib "io.github.gaelrenoux::tranzactio:4.1.0"
//> using lib "org.tpolecat::doobie-core:1.0.0-RC2"
//> using lib "org.tpolecat::doobie-postgres:1.0.0-RC2"

import zio._
import javax.sql.DataSource

import io.github.gaelrenoux.tranzactio.doobie._
import io.github.gaelrenoux.tranzactio.doobie.{ Database => TZDatabase }
import org.postgresql.ds.PGSimpleDataSource

object Database {
  val dsLayer: ULayer[DataSource] = ZLayer(ZIO.succeed(new PGSimpleDataSource()))
  val live: ULayer[TZDatabase]    = dsLayer >>> TZDatabase.fromDatasource
}

case class Service(database: TZDatabase) {
  def method: Task[Int] = ZIO.succeed(1)
}

object Service {
  def live: RLayer[TZDatabase, Service] = ZLayer.fromFunction(Service(_))
}

object Repro extends ZIOAppDefault {
  val run = ZIO.serviceWithZIO[Service](_.method.debug).provide(Service.live, Database.live)
}
  ```
  
  And this triggers the following error:
  
  ```
  timestamp=2022-12-07T08:44:57.396552Z level=ERROR thread=#zio-fiber-0 message="" cause="Exception in thread "zio-fiber-4" java.lang.Error: Defect in zio.ZEnvironment: Could not find DatabaseOps::ServiceOps[=transactor::Transactor[=λ %A → ZIO[-Any,+Throwable,+A]]] inside ZEnvironment(DatabaseServiceBase[=transactor$::Transactor[=λ %A → ZIO[-Any,+Throwable,+A]]] -> io.github.gaelrenoux.tranzactio.doobie.package$Database$$anon$4@6c558f81)
        at <empty>.Service.live(repro.scala:46)
        at <empty>.Repro.run(repro.scala:49)
        at <empty>.Repro.run(repro.scala:49)
        at <empty>.Repro.run(repro.scala:49)
        at <empty>.Repro.run(repro.scala:49)"
```
I'm pretty sure that TZDatabase is a DatabaseOps.ServiceOps[Connection] but there is an issue with tag, subtype and I really don't know how we could tackle that. This might be a ZIO bug as well (using 2.0.5).

I'm pretty confused with the types here, I don't know if there is something we could do to work around that. Also, when switching to scala 2.13 (easily changeable at the top of the file), everything works fine.

Scastie: https://scastie.scala-lang.org/rDWLHazKQqS2bulxpvP9Rw

**Comment by guersam** [comment]
It seems to happen since ZIO 2.0.3: https://github.com/zio/zio/issues/7506

**Comment by guersam** [comment]
> Does this mean that upgrading to ZIO 2.0.3 or more in tranzactio could solve this issue ?

I think so as it might be a backward runtime compatibility issue of izumi-reflect or ZIO.

If you copy and paste the layer construction code, it will work again even with ZIO >= 2.0.3

**Comment by ex0ns** [comment]
Oh yes, that could work, I tried it with that:

```scala
private final def fromConnectionSource(implicit trace: Trace): ZLayer[ConnectionSource, Nothing, TZDatabase] =
    ZLayer.fromFunction { (cs: ConnectionSource) =>
      new DatabaseServiceBase[Connection](cs) {
        override final def connectionFromJdbc(connection: => JdbcConnection)(implicit
          trace: Trace
        ): ZIO[Any, Nothing, Connection] =
          TZDatabase.connectionFromJdbc(connection)
      }
    }
  val step = (ConnectionSource.fromDatasource >>> fromConnectionSource)
```

However this only push the problem one step lower:
```
timestamp=2022-12-07T11:24:47.571049Z level=ERROR thread=#zio-fiber-0 message="" cause="Exception in thread "zio-fiber-4" java.lang.Error: Defect in zio.ZEnvironment: Could not find ConnectionSource::Service inside ZEnvironment(ConnectionSource$::DatasourceService -> io.github.gaelrenoux.tranzactio.ConnectionSource$DatasourceService@32ca8c42)
```

And the issue is that both instances of `ServiceBase` (namely `DatasourceService` and `SingleConnectionService`) are private, which forces me to do the following:

```scala
  val datasource: ULayer[DataSource] = ZLayer(ZIO.succeed(new PGSimpleDataSource()))

  class CustomDatasourceService(
    dataSource: DataSource,
    defaultErrorStrategies: ErrorStrategiesRef
  ) extends ServiceBase(defaultErrorStrategies) {

    override def getConnection(implicit trace: Trace): RIO[Any, JdbcConnection] = ZIO.attemptBlocking {
      dataSource.getConnection()
    }
  }
  private final def fromConnectionSource(implicit trace: Trace): ZLayer[ConnectionSource, Nothing, TZDatabase] =
    ZLayer.fromFunction { (cs: ConnectionSource) =>
      new DatabaseServiceBase[Connection](cs) {
        override final def connectionFromJdbc(connection: => JdbcConnection)(implicit
          trace: Trace
        ): ZIO[Any, Nothing, Connection] =
          TZDatabase.connectionFromJdbc(connection)
      }
    }
  val fromDataSource = ZLayer {
    for {
      source <- ZIO.service[DataSource]
    } yield new CustomDatasourceService(source, ErrorStrategies.Parent)
  }
  val live = datasource >>> (fromDataSource >>> fromConnectionSource)
  ```
  
  And now it works great I would never have found without your help. Thanks a lot for the tip, I will bury that somewhere in my code and remove it when the ecosystem has been upgraded to 2.0.3+.
  
 To avoid such issues in the future (even if they should be prevented at the ZIO level IMO), maybe we could provide some way of creating ZIO instead of ZLayer in tranzactio, so the user could manually wrap them if needed (I know this is a pretty niche usecase)

**Comment by narma** [comment]
From this [source](https://github.com/zio/zio/issues/7506#issuecomment-1506938237) to fix this need to update izumi-reflect in lib to 2.2.0 or above, recompile and republish artefact.

**Comment by gaelrenoux** [comment]
Hey, sorry for the long absence but well, life happened. Anyway, I've actually upgraded to the latest ZIO version (which required some tests to be fixed, 2.0.3 also broke compatibility in ZIOSpecAbstract). Feel free to reopen if this doesn't fix the issue.

**Comment by ex0ns** [comment]
No worries, thanks for taking care of it anyway ! 
We had a workaround that worked for the past few months, no rush on this side

The new version seems to work fine, however I had to explicitly include zio-interop-cats into my build

Thanks for the work
